openml/datasets/dataset.py

```diff
@@ -303,28 +347,98 @@ def decode_arff(fh):
             with io.open(filename, encoding='utf8') as fh:
                 return decode_arff(fh)
 
+    @staticmethod
+    def _convert_array_format(data, array_format, attribute_names):
+        """Convert a dataset to a given array format.
+
+        By default, the data are stored as a sparse matrix or a pandas
+        dataframe. One might be interested to get a pandas SparseDataFrame or a
+        NumPy array instead, respectively.
+        """
+        if array_format == "array" and not scipy.sparse.issparse(data):
+            # We encode the categories such that they are integer to be able
+            # to make a conversion to numeric for backward compatibility
+            def _encode_if_category(column):
+                if column.dtype.name == 'category':
+                    column = column.cat.codes
+                    column = column.apply(lambda x: np.nan if x == -1 else x)
+                return column
+            if data.ndim == 2:
+                for column_name in data.columns:
```

Comment:
I just briefly profiled the code and it looks like this for loop is the culprit of the slowdown. Replacing it by:
```
columns = {
    column_name: _encode_if_category(data.loc[:, column_name]) 
    for column_name in data.columns
}
                data = pd.DataFrame(columns)
```
increases the test speed again (while the unit tests are still passing). Could you please check (and replace if this is the case)?